tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc

```diff
@@ -335,6 +334,161 @@ void WarnIfBadDriverJITVersion() {
   });
 }
 
+PersistentCompilationCache::PersistentCompilationCache()
+{
+  cache_dir_ = GetDebugOptionsFromFlags().xla_gpu_persistent_cache_dir();
+  in_use_ = !cache_dir_.empty();
+  if (!in_use_) {
+    return;
+  }
+#if !defined(PLATFORM_POSIX)
+  // The current peristent cache design requires an atomic rename.
+  LOG(WARNING) << "XLA persistent cache is only supported on POSIX platforms.";
+  return;
+#endif
+  tensorflow::Env* env = tensorflow::Env::Default();
+  std::vector<string> files;
+  in_use_ = env->GetChildren(cache_dir_, &files).ok();
+  if (!in_use_) {
+    LOG(WARNING) << "Can't read XLA persistent cache directory \""
+                 << cache_dir_ << "\".";
+    return;
+  }
+  VLOG(2) << "Directory " << cache_dir_ << " is persistent cache.";
+  // Loop over the cache entries and store them in memory. Allow for failures. 
+  for (const auto& file : files) {
+    if (file[0] == '!') {
+      // Persistent cache temp file start with a '!'.
+      VLOG(3) << "Skipping temp file name \'" << file << "\".";
+      continue;
+    }
+    // The filename is the hash key.
+    int64 key;
+    if (!absl::SimpleAtoi(file, &key)) {
+      // Invalid filename. Filename must be an int64
+      LOG(WARNING) << "Skipping invalid entry name \"" << file << ".\"";
+      continue;
+    }
+    // Read the file, and store in cache on success.
+    std::string text;
+    string fullpath = tensorflow::io::JoinPath(cache_dir_, file);
+    if (!tensorflow::ReadFileToString(env, fullpath, &text).ok()) {
+      LOG(WARNING) << "Skippping entry \"" << fullpath << "\". Can't read it.";
+      continue;
+    }
+    // store PTX or cubin in memory cache
+    in_memory_cache_[key] = text;
+  }
+  VLOG(2) << "Persistent cache has " << in_memory_cache_.size()
+          << " entries.";
+  if (VLOG_IS_ON(3)) {
+    for (const auto &pair : in_memory_cache_) {
+      VLOG(3) << "  Entry: " << pair.first << ".";
+    }
+  }
+}
+
+constexpr const int64 PersistentCompilationCache::kPtxHash;
+
+int64 PersistentCompilationCache::CreateKey(
```

Comment:
What happens in case of hash collisions?  Or is that very unlikely?

Reply:
I use the same hash function as in xla_compilation_cache.cc. Collisions better be unlikely!

Reply:
I may be misreading the code, but here it seems like you assume if the hashes of two inputs match then they are the same input.  Is that correct?  I don't think xla_compilation_cache.cc makes that assumption.

Reply:
You are correct. I'm not knowledgeable on hash functions in general to know how good they are to generate unique hashes. In case a duplicate key is generated, that would indeed be bad. Since the LLVM IR is rather sizeable, it should be unlikely.

Reply:
I think to be sound you need to store the input IR as well as the output IR, and compare the input IRs in case the hashes match (just like a standard hashtable does).

Reply:
will do